File: src/emulator/peripherals/accelerometer/accelerometer.ts
import {Memory} from "../../memory/memory.ts"
import {BoardComponent} from "../../board/board-component.ts"
import {Board} from "../../board/board.ts"
import {PeripheralComponent} from "../peripheral-component.ts";
import {Ssu, ssuFlags} from "../../ssu/ssu.ts";

const RAM_SIZE = 0x7F

export const accelerometerState = {
    address: 0,
    data: 1
}

export class Accelerometer extends PeripheralComponent {
    memory: Memory

    state: number = accelerometerState.address
    address: number = 0
    offset: number = 0

    constructor(board: Board) {
        super(board)

        this.memory = Memory.fromSize(RAM_SIZE)
    }

    override transmitAndReceive(ssu: Ssu) {
        switch (this.state) {
            case accelerometerState.address:
                this.address = ssu.transmit
                this.offset = 0
                this.state = accelerometerState.data
                ssu.status |= ssuFlags.status.RECEIVE_FULL
                break
            case accelerometerState.data:
                ssu.receive = this.memory.readByte(this.address + this.offset)
                this.offset++
                ssu.status |= ssuFlags.status.RECEIVE_FULL
                ssu.status |= ssuFlags.status.TRANSMIT_EMPTY
                ssu.status |= ssuFlags.status.TRANSMIT_END

                break
        }
    }

    override transmit(ssu: Ssu) {
        switch (this.state) {
            case accelerometerState.address:
                this.address = ssu.transmit
                this.state = accelerometerState.data
                break
            case accelerometerState.data:
                this.board.accelerometer.memory.writeByte(this.address, ssu.transmit)
                ssu.status |= ssuFlags.status.TRANSMIT_EMPTY
                ssu.status |= ssuFlags.status.TRANSMIT_END

                break
        }
    }
}

File: PocketWalker/Emulator/Peripherals/Accelerometer/Accelerometer.cpp
#include "Accelerometer.h"

void Accelerometer::TransmitAndReceive(Ssu* ssu)
{
    switch (state)
    {
    case GettingAddress:
        {
            address = ssu->transmit;
            offset = 0;
            state = GettingData;

            ssu->status |= SsuFlags::Status::RECEIVE_FULL;
            break;
        }
    case GettingData:
        {
            ssu->receive = memory->ReadByte(address);
            offset++;

            ssu->status |= SsuFlags::Status::RECEIVE_FULL;
            ssu->status |= SsuFlags::Status::TRANSMIT_EMPTY;
            ssu->status |= SsuFlags::Status::TRANSMIT_END;
            break;
        }
    }
}

void Accelerometer::Transmit(Ssu* ssu)
{
    switch (state)
    {
    case GettingAddress:
        {
            address = ssu->transmit;
            state = GettingData;
            break;
        }
    case GettingData:
        {
            memory->WriteByte(address, ssu->transmit);

            ssu->status |= SsuFlags::Status::TRANSMIT_EMPTY;
            ssu->status |= SsuFlags::Status::TRANSMIT_END;
            break;
        }
    }
}

void Accelerometer::Reset()
{
    state = GettingAddress;
    offset = 0;
}
